


#' Make a blacklist for genomic regions
#' 
#' Produce a blacklist of genomic regions with a high ratio of duplicate to unique reads. This blacklist can be used to exclude reads for analysis in \code{\link{Aneufinder}}, \code{\link{bam2GRanges}} and \code{\link{bed2GRanges}}. This function produces a pre-blacklist which has to manually be filtered with a sensible cutoff. See the examples section for details.
#' 
#' @param files A character vector of either BAM or BED files.
#' @param bins A list with one \code{\link{GRanges}} with binned read counts generated by \code{\link{fixedWidthBins}}.
#' @inheritParams bed2GRanges
#' @inheritParams binReads
#' @return A \code{\link{GRanges}} with the same coordinates as \code{bins} with metadata columns ratio, duplicated counts and deduplicated counts.
#' @importFrom S4Vectors DataFrame
#' @export
#'@examples
#'## Get an example BAM file with single-cell-sequencing reads
#'bamfile <- system.file("extdata", "BB150803_IV_074.bam", package="AneuFinderData")
#'## Prepare the blacklist
#'bins <- fixedWidthBins(assembly='mm10', binsizes=1e6, chromosome.format='NCBI')
#'pre.blacklist <- blacklist(bamfile, bins=bins)
#'## Plot a histogram to decide on a sensible cutoff
#'qplot(pre.blacklist$ratio, binwidth=0.1)
#'## Make the blacklist with cutoff = 1.9
#'blacklist <- pre.blacklist[pre.blacklist$ratio > 1.9]
blacklist <- function(files, assembly, bins, min.mapq=10, pairedEndReads=FALSE) {
	
	## Input checks ##
	if (length(bins) > 1 & is.list(bins)) {
		stop("argument 'bins' must only contain one element")
	}
	if (class(bins) == 'GRanges') {
		bins <- list(bins)
		names(bins) <- width(bins[[1]])[1]
	}
	## Bin the reads with and without duplicates kept
	dupcounts.matrix <- array(dim=c(length(bins[[1]]), length(files)), dimnames=list(bin=1:length(bins[[1]]), file=basename(files)))
	counts.matrix <- dupcounts.matrix
	for (file in files) {
		dup.binned <- binReads(file, assembly = assembly, min.mapq = min.mapq, pairedEndReads = pairedEndReads, binsizes = NULL, bins = bins, chromosomes = seqlevels(bins[[1]]), calc.complexity = FALSE, remove.duplicate.reads = FALSE)
		dupcounts.matrix[,basename(file)] <- dup.binned[[1]]$counts
		binned <- binReads(file, assembly = assembly, min.mapq = min.mapq, pairedEndReads = pairedEndReads, binsizes = NULL, bins = bins, chromosomes = seqlevels(bins[[1]]), calc.complexity = FALSE, remove.duplicate.reads = TRUE)
		counts.matrix[,basename(file)] <- binned[[1]]$counts
	}
	dupcounts <- rowSums(dupcounts.matrix)
	counts <- rowSums(counts.matrix)
	ratio <- dupcounts / counts
	ratio[is.na(ratio)] <- 1
	
	pre.blacklist <- bins[[1]]
	mcols(pre.blacklist) <- S4Vectors::DataFrame(ratio, dupcounts, counts)
	
	return(pre.blacklist)
	
}
